api/bootstrap/kubeadm/v1beta2/kubeadm_types.go

```diff
@@ -424,7 +424,7 @@ type BootstrapToken struct {
 	// token is used for establishing bidirectional trust between nodes and control-planes.
 	// Used for joining nodes in the cluster.
 	// +required
-	Token *BootstrapTokenString `json:"token"`
+	Token BootstrapTokenString `json:"token,omitempty,omitzero"`
```

Comment:
q about the similar changes recently. for serialization that seems fine, but golang and api importers it's still valuable to have pointer fields and distinguishing nil and empty, correct?

Reply:
I would say it's a trade-off.

Yes it's easier to check if it is set or not if it is a pointer.

In many cases we wanted to rather check if specific fields in a struct field were set or not. So having a pointer let to a *lot* of additional nil checks everywhere.

In cases where we had a lot of cases where we wanted to check for set or not we introduced `IsDefined` methods, like this:
```go
// IsDefined returns true if the LocalEtcd is defined.
func (r *LocalEtcd) IsDefined() bool {
	return !reflect.DeepEqual(r, &LocalEtcd{})
}


// IsDefined returns true if the ContractVersionedObjectReference is set.
func (r *ContractVersionedObjectReference) IsDefined() bool {
	if r == nil {
		return false
	}
	return r.Kind != "" || r.Name != "" || r.APIGroup != ""
}

```

Reply:
I think you've captured it well @sbueringer , in most cases, APIs aren't switching their behaviour based on if the struct is present or not, but in fact if fields within the struct have non-zero values. So in a lot of cases, the `!= nil` check just gates some other condition. I think the `IsDefined` pattern that Stefan has introduced should make it still possible for folks to work out if/if not a struct is set